cub/cub/device/dispatch/kernels/transform.cuh

```diff
@@ -609,14 +610,44 @@ _CCCL_DEVICE void bulk_copy_maybe_unaligned(
   }
 }
 
+template <int Alignment>
+_CCCL_DEVICE auto align_shared_ptr_now(char* p) -> char*
+{
+  // shared memory addresses are 32-bit, so convert and align it
+  uint32_t smem32 = __cvta_generic_to_shared(p);
+  smem32          = ::cuda::round_up(smem32, Alignment);
+  // pretend to NVVM that the 32-bit pointer is modified. this is required to avoid constant propagation from pulling
+  // the smem32 definition into loops and branches in subsequent code. thus, the pointer alignment code remains at the
+  // start of the kernel.
+  asm("" : "+r"(smem32));
+  // convert back to a generic pointer for C++ code
+  return static_cast<char*>(__cvta_shared_to_generic(smem32));
```

Comment:
Nitpick, given that the alignment is a compile time constant we could use `_CCCL_BUILTIN_ASSUME_ALIGNED`
```suggestion
  return _CCCL_BUILTIN_ASSUME_ALIGNED(static_cast<char*>(__cvta_shared_to_generic(smem32)), Alignment);
```